src/core/geometry/qgspolyhedralsurface.cpp

```diff
@@ -0,0 +1,997 @@
+      clear();
+      return false;
+    }
+  }
+
+  return true;
+}
+
+QgsBox3D QgsPolyhedralSurface::calculateBoundingBox3D() const
+{
+  if ( mPatches.empty() )
+  {
+    return QgsBox3D();
+  }
+
+  QgsBox3D bbox = mPatches.at( 0 )->boundingBox3D();
+  for ( int i = 1; i < mPatches.size(); ++i )
+  {
+    QgsBox3D polygonBox = mPatches.at( i )->boundingBox3D();
+    bbox.combineWith( polygonBox );
+  }
+  return bbox;
+}
+
+int QgsPolyhedralSurface::wkbSize( QgsAbstractGeometry::WkbFlags flags ) const
+{
+  int binarySize = sizeof( char ) + sizeof( quint32 ) + sizeof( quint32 );
+  for ( const QgsPolygon *patch : mPatches )
+  {
+    binarySize += patch->wkbSize( flags );
+  }
+  return binarySize;
+}
+
+QByteArray QgsPolyhedralSurface::asWkb( WkbFlags flags ) const
+{
+  QByteArray wkbArray;
+  wkbArray.resize( QgsPolyhedralSurface::wkbSize( flags ) );
+  QgsWkbPtr wkb( wkbArray );
+  wkb << static_cast<char>( QgsApplication::endian() );
+  wkb << static_cast<quint32>( wkbType() );
+  wkb << static_cast<quint32>( mPatches.size() );
+  for ( const QgsPolygon *patch : mPatches )
+  {
+    wkb << patch->asWkb( flags );
+  }
+  return wkbArray;
+}
+
+QString QgsPolyhedralSurface::asWkt( int precision ) const
+{
+  QString wkt = wktTypeStr();
+
+  if ( isEmpty() )
+    wkt += QLatin1String( " EMPTY" );
+  else
+  {
+    wkt += QLatin1String( " (" );
+    for ( const QgsPolygon *patch : mPatches )
+    {
+      QString childWkt = patch->asWkt( precision );
+      if ( qgsgeometry_cast<const QgsPolygon *>( patch ) )
+      {
+        // Type names of linear geometries are omitted
+        childWkt = childWkt.mid( childWkt.indexOf( '(' ) );
+      }
+      wkt += childWkt + ',';
+    }
+    if ( wkt.endsWith( ',' ) )
+    {
+      wkt.chop( 1 ); // Remove last ','
+    }
+    wkt += ')';
+  }
+  return wkt;
+}
+
+QDomElement QgsPolyhedralSurface::asGml2( QDomDocument &, int, const QString &, const AxisOrder ) const
+{
+  QgsDebugError( QStringLiteral( "gml version 2 does not support PolyhedralSurface geometry" ) );
+  return QDomElement();
+}
+
+QDomElement QgsPolyhedralSurface::asGml3( QDomDocument &doc, int precision, const QString &ns, const QgsAbstractGeometry::AxisOrder axisOrder ) const
+{
+  QDomElement elemPolyhedralSurface = doc.createElementNS( ns, QStringLiteral( "PolyhedralSurface" ) );
+
+  if ( isEmpty() )
+    return elemPolyhedralSurface;
+
+  QDomElement elemPolygonPatches = doc.createElementNS( ns, QStringLiteral( "polygonPatches" ) );
+
+  for ( QgsPolygon *patch : mPatches )
+  {
+    QDomElement elemPolygonPatch = patch->asGml3( doc, precision, ns, axisOrder );
+    elemPolygonPatch.setTagName( "PolygonPatch" );
+
+    elemPolygonPatches.appendChild( elemPolygonPatch );
+  }
+  elemPolyhedralSurface.appendChild( elemPolygonPatches );
+
+  return elemPolyhedralSurface;
+}
+
+json QgsPolyhedralSurface::asJsonObject( int ) const
+{
+  QgsDebugError( QStringLiteral( "geojson format does not support PolyhedralSurface geometry" ) );
```

Comment:
Shouldn't we treat this as a GeoJSON multipolygon then?

Reply:
Done.